build_tools/cmake/iree_setup_toolchain.cmake

```diff
@@ -15,168 +15,178 @@ function(iree_append_to_lists VALUE)
+  # use that.
+  if(IREE_ENABLE_SPLIT_DWARF)
+    check_cxx_compiler_flag(-gsplit-dwarf IREE_SUPPORTS_SPLIT_DWARF)
+    if(IREE_SUPPORTS_SPLIT_DWARF)
+      # Also add -ggnu-pubnames for compilation because it links faster and lld
+      # doesn't do the slow path without it.
+      iree_append_to_lists(" -gsplit-dwarf -ggnu-pubnames"
+        CMAKE_C_FLAGS_DEBUG
+        CMAKE_CXX_FLAGS_DEBUG
+        CMAKE_C_FLAGS_RELWITHDEBINFO
+        CMAKE_CXX_FLAGS_RELWITHDEBINFO
+      )
+    endif()
+    check_linker_flag(CXX "-Wl,--gdb-index" IREE_SUPPORTS_GDB_INDEX)
+    if(IREE_SUPPORTS_GDB_INDEX)
+      message(STATUS "Enabling gdb-index (binaries with debug info are not relocatable)")
+      iree_append_to_lists(" -Wl,--gdb-index"
+        CMAKE_EXE_LINKER_FLAGS_DEBUG
+        CMAKE_MODULE_LINKER_FLAGS_DEBUG
+        CMAKE_SHARED_LINKER_FLAGS_DEBUG
+        CMAKE_EXE_LINKER_FLAGS_RELWITHDEBINFO
+        CMAKE_MODULE_LINKER_FLAGS_RELWITHDEBINFO
+        CMAKE_SHARED_LINKER_FLAGS_RELWITHDEBINFO
+      )
+    endif()
   endif()
-endif()
-
-# Thin archives makes static archives that only link to backing object files
-# instead of embedding them. This makes them non-relocatable but is almost
-# always the right thing outside of certain deployment/packaging scenarios.
-if(IREE_ENABLE_THIN_ARCHIVES)
-  execute_process(COMMAND ${CMAKE_AR} -V OUTPUT_VARIABLE IREE_AR_VERSION)
-  if ("${IREE_AR_VERSION}" MATCHES "^GNU ar|LLVM")
-    message(STATUS "Enabling thin archives (static libraries will not be relocatable)")
-    set(CMAKE_C_ARCHIVE_APPEND "<CMAKE_AR> qT <TARGET> <LINK_FLAGS> <OBJECTS>")
-    set(CMAKE_CXX_ARCHIVE_APPEND "<CMAKE_AR> qT <TARGET> <LINK_FLAGS> <OBJECTS>")
-    set(CMAKE_C_ARCHIVE_CREATE "<CMAKE_AR> crT <TARGET> <LINK_FLAGS> <OBJECTS>")
-    set(CMAKE_CXX_ARCHIVE_CREATE "<CMAKE_AR> crT <TARGET> <LINK_FLAGS> <OBJECTS>")
-  else()
-    message(WARNING "Thin archives requested but not supported by ar")
+
+  # Thin archives makes static archives that only link to backing object files
+  # instead of embedding them. This makes them non-relocatable but is almost
+  # always the right thing outside of certain deployment/packaging scenarios.
+  if(IREE_ENABLE_THIN_ARCHIVES)
+    execute_process(COMMAND ${CMAKE_AR} -V OUTPUT_VARIABLE IREE_AR_VERSION)
+    if ("${IREE_AR_VERSION}" MATCHES "^GNU ar|LLVM")
+      message(STATUS "Enabling thin archives (static libraries will not be relocatable)")
+      set(CMAKE_C_ARCHIVE_APPEND "<CMAKE_AR> qT <TARGET> <LINK_FLAGS> <OBJECTS>")
+      set(CMAKE_CXX_ARCHIVE_APPEND "<CMAKE_AR> qT <TARGET> <LINK_FLAGS> <OBJECTS>")
+      set(CMAKE_C_ARCHIVE_CREATE "<CMAKE_AR> crT <TARGET> <LINK_FLAGS> <OBJECTS>")
+      set(CMAKE_CXX_ARCHIVE_CREATE "<CMAKE_AR> crT <TARGET> <LINK_FLAGS> <OBJECTS>")
+    else()
+      message(WARNING "Thin archives requested but not supported by ar")
+    endif()
   endif()
-endif()
+endmacro()
+
+iree_setup_toolchain()
```

Comment:
Don't we want this defined in a separate file than it's called? Otherwise, how is this functionally difficult for downstream projects?

Reply:
Macros (and functions) are global (as in global-global). So if you `add_subdirectory(path_to_iree)` in a wrapping project, they will get access to this macro in the same way they get access to `iree_cc_library`, et al.

And yes, no changes were made to what is now the macro except indenting the body.